{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Concordance between resfinder and resistance/susceptible status"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n",
      "The dotenv extension is already loaded. To reload it, use:\n",
      "  %reload_ext dotenv\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the \"autoreload\" extension\n",
    "%load_ext autoreload\n",
    "\n",
    "# always reload modules marked with \"%aimport\"\n",
    "%autoreload 1\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import dotenv\n",
    "\n",
    "from sklearn.externals import joblib\n",
    "\n",
    "# add the 'src' directory as one where we can import modules\n",
    "project_dir = os.path.join(os.getcwd(), os.pardir)\n",
    "src_dir = os.path.join(project_dir, 'src')\n",
    "sys.path.append(src_dir)\n",
    "\n",
    "# import my method from the source code\n",
    "%aimport data.tools\n",
    "\n",
    "# load env\n",
    "%load_ext dotenv\n",
    "dotenv_path = os.path.join(project_dir, '.env')\n",
    "dotenv.load_dotenv(dotenv_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load master Xlsx sheet\n",
    "import pandas as pd\n",
    "amrfile = os.path.join(os.environ.get('PRDATA'),'raw','GenotypicAMR_Master.xlsx')\n",
    "amrdf = pd.read_excel(amrfile)\n",
    "\n",
    "amrdf = amrdf.replace(r'\\s+', np.nan, regex=True)\n",
    "amrdf = amrdf.replace(r'-', np.nan, regex=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "------------------------------\n",
      "blaCARB-2|M69058:\n",
      "accuracy: 0.6381811054488435\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.78      1602\n",
      "        1.0       0.96      0.03      0.06       949\n",
      "\n",
      "avg / total       0.76      0.64      0.51      2551\n",
      "\n",
      "TN (susceptible): 1601\n",
      "FN (resistant predicted as susceptible): 922\n",
      "FP (susceptible predicted as resistant): 1\n",
      "TP (resistant): 27\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaCMY-17|AY513266:\n",
      "accuracy: 0.6283810270482164\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.00       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 948\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 1\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaCMY-2|X91840:\n",
      "accuracy: 0.7840062720501764\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.74      1.00      0.85      1602\n",
      "        1.0       1.00      0.42      0.59       949\n",
      "\n",
      "avg / total       0.84      0.78      0.76      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 551\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 398\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaCMY-42|HM146927:\n",
      "accuracy: 0.6295570364562917\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.01       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 945\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 4\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaCMY-4|AF420597:\n",
      "accuracy: 0.6283810270482164\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.00       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 948\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 1\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaCMY-6|AJ011293:\n",
      "accuracy: 0.6287730301842415\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.00       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 947\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 2\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaCTX-M-14b|DQ359215:\n",
      "accuracy: 0.6283810270482164\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.00       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 948\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 1\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaCTX-M-14|AF252622:\n",
      "accuracy: 0.630733045864367\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.01      0.01       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 942\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 7\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaCTX-M-15|DQ302097:\n",
      "accuracy: 0.6707173657389259\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.66      1.00      0.79      1602\n",
      "        1.0       1.00      0.11      0.21       949\n",
      "\n",
      "avg / total       0.78      0.67      0.57      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 840\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 109\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaCTX-M-1|DQ915955:\n",
      "accuracy: 0.6287730301842415\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.00       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 947\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 2\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaCTX-M-27|EU916273:\n",
      "accuracy: 0.6358290866326931\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.78      1602\n",
      "        1.0       1.00      0.02      0.04       949\n",
      "\n",
      "avg / total       0.77      0.64      0.50      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 929\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 20\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaCTX-M-55|GQ456159:\n",
      "accuracy: 0.6330850646805174\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.01      0.03       949\n",
      "\n",
      "avg / total       0.77      0.63      0.50      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 936\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 13\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaCTX-M-65|GQ456158:\n",
      "accuracy: 0.6287730301842415\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.00       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 947\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 2\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaCTX-M-8|AF189721:\n",
      "accuracy: 0.6283810270482164\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.00       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 948\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 1\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaDHA-1|Y16410:\n",
      "accuracy: 0.6287730301842415\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.00       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 947\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 2\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaFOX-5|AY007369:\n",
      "accuracy: 0.6283810270482164\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.00       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 948\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 1\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaKPC-2|AY034847:\n",
      "accuracy: 0.6291650333202665\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.01       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 946\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 3\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaKPC-3|HM769262:\n",
      "accuracy: 0.6319090552724422\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.01      0.02       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 939\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 10\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaKPC-4|FJ473382:\n",
      "accuracy: 0.6283810270482164\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.00       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 948\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 1\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaNDM-1|FN396876:\n",
      "accuracy: 0.6287730301842415\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.00       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 947\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 2\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaNDM-5|JN104597:\n",
      "accuracy: 0.6295570364562917\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.01       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 945\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 4\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaNDM-6|JN967644:\n",
      "accuracy: 0.6283810270482164\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.00       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 948\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 1\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaNDM-7|JX262694:\n",
      "accuracy: 0.6283810270482164\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.00       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 948\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 1\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaOXA-181|HM992946:\n",
      "accuracy: 0.6283810270482164\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.00       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 948\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 1\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaOXA-1|J02967:\n",
      "accuracy: 0.6597412779302234\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.65      1.00      0.79      1602\n",
      "        1.0       1.00      0.09      0.16       949\n",
      "\n",
      "avg / total       0.78      0.66      0.55      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 868\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 81\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaOXA-2|DQ310703:\n",
      "accuracy: 0.6291650333202665\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.01       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 946\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 3\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaOXA-9|JF703130:\n",
      "accuracy: 0.6295570364562917\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.01       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 945\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 4\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaSHV-105|FJ194944:\n",
      "accuracy: 0.6283810270482164\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.00       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 948\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 1\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaSHV-12|AF462395:\n",
      "accuracy: 0.6299490395923167\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.01      0.01       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 944\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 5\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaSHV-2|GU064394:\n",
      "accuracy: 0.6283810270482164\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.00       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 948\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 1\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaTEM-199|JX050178:\n",
      "accuracy: 0.6287730301842415\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.00       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 947\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 2\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaTEM-1A|HM749966:\n",
      "accuracy: 0.6405331242649941\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.64      1.00      0.78      1602\n",
      "        1.0       1.00      0.03      0.07       949\n",
      "\n",
      "avg / total       0.77      0.64      0.51      2551\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 917\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 32\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaTEM-1B|JF910132:\n",
      "accuracy: 0.7745981967855743\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.74      1.00      0.85      1602\n",
      "        1.0       1.00      0.40      0.57       949\n",
      "\n",
      "avg / total       0.83      0.77      0.74      2551\n",
      "\n",
      "TN (susceptible): 1601\n",
      "FN (resistant predicted as susceptible): 574\n",
      "FP (susceptible predicted as resistant): 1\n",
      "TP (resistant): 375\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaTEM-1C|FJ560503:\n",
      "accuracy: 0.6342610740885927\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.02      0.03       949\n",
      "\n",
      "avg / total       0.77      0.63      0.50      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 933\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 16\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaTEM-1D|AF188200:\n",
      "accuracy: 0.6283810270482164\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.00       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 948\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 1\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaTEM-206|KC783461:\n",
      "accuracy: 0.6295570364562917\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.01       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 945\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 4\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaTEM-33|GU371926:\n",
      "accuracy: 0.6287730301842415\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.00       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 947\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 2\n",
      "\n",
      "\n",
      "------------------------------\n",
      "blaTEM-52B|AF027199:\n",
      "accuracy: 0.6283810270482164\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.63      1.00      0.77      1602\n",
      "        1.0       1.00      0.00      0.00       949\n",
      "\n",
      "avg / total       0.77      0.63      0.49      2551\n",
      "\n",
      "TN (susceptible): 1602\n",
      "FN (resistant predicted as susceptible): 948\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 1\n"
     ]
    }
   ],
   "source": [
    "# Compute accuracy of a single gene predicting resistant/susceptible\n",
    "from sklearn.metrics import accuracy_score, classification_report, confusion_matrix\n",
    "import numpy as np\n",
    "\n",
    "# Beta lactamases\n",
    "bladf = amrdf.filter(regex=(\"bla*\"))\n",
    "\n",
    "y_true = amrdf['R_AMP'].as_matrix().astype(float)\n",
    "y_true[y_true == 0.5] = 0 # Set intermediate to susceptible\n",
    "missing = np.isnan(y_true)\n",
    "\n",
    "# Compute accuracy for each individual beta-lactamase\n",
    "for g in bladf:\n",
    "    y_pred = bladf[g].as_matrix().astype(float)\n",
    "    missing2 = np.isnan(y_pred)\n",
    "    \n",
    "    y_true2 = y_true[~missing & ~missing2]\n",
    "    y_pred2 = y_pred[~missing & ~missing2]\n",
    "    \n",
    "    print()\n",
    "    print(\"\\n------------------------------\\n\"+g+\":\")\n",
    "    print(\"accuracy: \"+str(accuracy_score(y_true2,y_pred2)))\n",
    "    print(classification_report(y_true2, y_pred2))\n",
    "    tn, fp, fn, tp = confusion_matrix(y_true2, y_pred2).ravel()\n",
    "    print(\"TN (susceptible): \"+str(tn))\n",
    "    print(\"FN (resistant predicted as susceptible): \"+str(fn))\n",
    "    print(\"FP (susceptible predicted as resistant): \"+str(fp))\n",
    "    print(\"TP (resistant): \"+str(tp))\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ALL beta-lactamase:\n",
      "\n",
      "------------------------------\n",
      "R_AMC:\n",
      "accuracy: 0.8229984301412873\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.99      0.79      0.88      2071\n",
      "        1.0       0.51      0.98      0.67       477\n",
      "\n",
      "avg / total       0.90      0.82      0.84      2548\n",
      "\n",
      "TN (susceptible): 1629\n",
      "FN (resistant predicted as susceptible): 9\n",
      "FP (susceptible predicted as resistant): 442\n",
      "TP (resistant): 468\n",
      "\n",
      "------------------------------\n",
      "R_AMP:\n",
      "accuracy: 0.9843198745589965\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.98      1.00      0.99      1602\n",
      "        1.0       1.00      0.96      0.98       949\n",
      "\n",
      "avg / total       0.98      0.98      0.98      2551\n",
      "\n",
      "TN (susceptible): 1600\n",
      "FN (resistant predicted as susceptible): 38\n",
      "FP (susceptible predicted as resistant): 2\n",
      "TP (resistant): 911\n",
      "\n",
      "------------------------------\n",
      "R_TIO:\n",
      "accuracy: 0.8559794256322332\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       1.00      0.83      0.91      1956\n",
      "        1.0       0.53      0.99      0.69       377\n",
      "\n",
      "avg / total       0.92      0.86      0.87      2333\n",
      "\n",
      "TN (susceptible): 1625\n",
      "FN (resistant predicted as susceptible): 5\n",
      "FP (susceptible predicted as resistant): 331\n",
      "TP (resistant): 372\n",
      "\n",
      "------------------------------\n",
      "R_FOX:\n",
      "accuracy: 0.7960815673730508\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       1.00      0.76      0.86      2134\n",
      "        1.0       0.42      0.98      0.59       367\n",
      "\n",
      "avg / total       0.91      0.80      0.82      2501\n",
      "\n",
      "TN (susceptible): 1630\n",
      "FN (resistant predicted as susceptible): 6\n",
      "FP (susceptible predicted as resistant): 504\n",
      "TP (resistant): 361\n",
      "\n",
      "------------------------------\n",
      "R_CRO:\n",
      "accuracy: 0.8635829086632693\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "        0.0       0.99      0.83      0.90      1968\n",
      "        1.0       0.63      0.98      0.77       583\n",
      "\n",
      "avg / total       0.91      0.86      0.87      2551\n",
      "\n",
      "TN (susceptible): 1629\n",
      "FN (resistant predicted as susceptible): 9\n",
      "FP (susceptible predicted as resistant): 339\n",
      "TP (resistant): 574\n"
     ]
    }
   ],
   "source": [
    "# Compute accuracy for any beta-lactamase\n",
    "anybla = np.zeros(bladf.shape[0])\n",
    "for g in bladf:\n",
    "    y = bladf[g].as_matrix().astype(float)\n",
    "    anybla = np.logical_or(anybla, y)\n",
    "\n",
    "anybla = anybla.astype(float)\n",
    "missing2 = np.isnan(anybla)\n",
    "\n",
    "\n",
    "\n",
    "print('ALL beta-lactamase:')\n",
    "\n",
    "for g in ['R_AMC', 'R_AMP', 'R_TIO', 'R_FOX', 'R_CRO']:\n",
    "    y_true = amrdf[g].as_matrix().astype(float)\n",
    "    y_true[y_true == 0.5] = 0 # Set intermediate to susceptible\n",
    "    missing = np.isnan(y_true)\n",
    "    y_true2 = y_true[~missing & ~missing2]\n",
    "    y_pred = anybla[~missing & ~missing2]\n",
    "    print()\n",
    "    print(\"------------------------------\\n\"+g+\":\")\n",
    "    print(\"accuracy: \"+str(accuracy_score(y_true2,y_pred)))\n",
    "    print(classification_report(y_true2, y_pred))\n",
    "    tn, fp, fn, tp = confusion_matrix(y_true2, y_pred).ravel()\n",
    "    print(\"TN (susceptible): \"+str(tn))\n",
    "    print(\"FN (resistant predicted as susceptible): \"+str(fn))\n",
    "    print(\"FP (susceptible predicted as resistant): \"+str(fp))\n",
    "    print(\"TP (resistant): \"+str(tp))\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 443,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "DRUG: R_AMC\n",
      "\n",
      "------------------------------\n",
      "DecisionTree Set blaCMY-2|X91840,blaCMY-42|HM146927,blaKPC-2|AY034847,blaKPC-3|HM769262,blaNDM-5|JN104597,blaOXA-1|J02967\n",
      "\n",
      "accuracy: 0.967032967032967\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "          0       0.98      0.98      0.98      2071\n",
      "          1       0.90      0.93      0.91       477\n",
      "\n",
      "avg / total       0.97      0.97      0.97      2548\n",
      "\n",
      "TN (susceptible): 2022\n",
      "FN (resistant predicted as susceptible): 35\n",
      "FP (susceptible predicted as resistant): 49\n",
      "TP (resistant): 442\n",
      "\n",
      "DRUG: R_AMP\n",
      "\n",
      "------------------------------\n",
      "DecisionTree Set blaCARB-2|M69058,blaCMY-2|X91840,blaCTX-M-14|AF252622,blaCTX-M-15|DQ302097,blaCTX-M-1|DQ915955,blaCTX-M-27|EU916273,blaCTX-M-55|GQ456159,blaKPC-2|AY034847,blaOXA-1|J02967,blaOXA-2|DQ310703,blaTEM-1A|HM749966,blaTEM-1B|JF910132,blaTEM-1C|FJ560503,blaTEM-33|GU371926\n",
      "\n",
      "accuracy: 0.9815758526068209\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "          0       0.97      1.00      0.99      1602\n",
      "          1       1.00      0.95      0.97       949\n",
      "\n",
      "avg / total       0.98      0.98      0.98      2551\n",
      "\n",
      "TN (susceptible): 1600\n",
      "FN (resistant predicted as susceptible): 45\n",
      "FP (susceptible predicted as resistant): 2\n",
      "TP (resistant): 904\n",
      "\n",
      "DRUG: R_TIO\n",
      "\n",
      "------------------------------\n",
      "DecisionTree Set blaCMY-2|X91840\n",
      "\n",
      "accuracy: 0.992284612087441\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "          0       0.99      1.00      1.00      1956\n",
      "          1       0.98      0.97      0.98       377\n",
      "\n",
      "avg / total       0.99      0.99      0.99      2333\n",
      "\n",
      "TN (susceptible): 1950\n",
      "FN (resistant predicted as susceptible): 12\n",
      "FP (susceptible predicted as resistant): 6\n",
      "TP (resistant): 365\n",
      "\n",
      "DRUG: R_FOX\n",
      "\n",
      "------------------------------\n",
      "DecisionTree Set blaCMY-2|X91840,blaCMY-42|HM146927,blaCTX-M-14|AF252622,blaKPC-2|AY034847,blaKPC-3|HM769262,blaNDM-5|JN104597,blaOXA-1|J02967\n",
      "\n",
      "accuracy: 0.9500199920031988\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "          0       0.99      0.95      0.97      2134\n",
      "          1       0.76      0.96      0.85       367\n",
      "\n",
      "avg / total       0.96      0.95      0.95      2501\n",
      "\n",
      "TN (susceptible): 2023\n",
      "FN (resistant predicted as susceptible): 14\n",
      "FP (susceptible predicted as resistant): 111\n",
      "TP (resistant): 353\n",
      "\n",
      "DRUG: R_CRO\n",
      "\n",
      "------------------------------\n",
      "DecisionTree Set blaCMY-2|X91840,blaCTX-M-14|AF252622,blaCTX-M-15|DQ302097,blaCTX-M-1|DQ915955,blaCTX-M-27|EU916273,blaCTX-M-55|GQ456159,blaCTX-M-65|GQ456158,blaKPC-2|AY034847,blaKPC-3|HM769262,blaSHV-12|AF462395,blaTEM-33|GU371926\n",
      "\n",
      "accuracy: 0.9898079184633477\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "          0       0.99      1.00      0.99      1968\n",
      "          1       0.99      0.96      0.98       583\n",
      "\n",
      "avg / total       0.99      0.99      0.99      2551\n",
      "\n",
      "TN (susceptible): 1965\n",
      "FN (resistant predicted as susceptible): 23\n",
      "FP (susceptible predicted as resistant): 3\n",
      "TP (resistant): 560\n",
      "\n",
      "DRUG: phenotypic_aztreonam\n",
      "\n",
      "------------------------------\n",
      "DecisionTree Set blaCMY-2|X91840,blaCTX-M-15|DQ302097,blaCTX-M-55|GQ456159,blaKPC-2|AY034847,blaKPC-3|HM769262\n",
      "\n",
      "accuracy: 0.8584070796460177\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "          0       0.60      0.71      0.65        21\n",
      "          1       0.93      0.89      0.91        92\n",
      "\n",
      "avg / total       0.87      0.86      0.86       113\n",
      "\n",
      "TN (susceptible): 15\n",
      "FN (resistant predicted as susceptible): 10\n",
      "FP (susceptible predicted as resistant): 6\n",
      "TP (resistant): 82\n",
      "\n",
      "DRUG: phenotypic_cefazolin\n",
      "\n",
      "------------------------------\n",
      "DecisionTree Set blaCMY-2|X91840,blaCMY-4|AF420597,blaCTX-M-14b|DQ359215,blaCTX-M-14|AF252622,blaCTX-M-15|DQ302097,blaCTX-M-1|DQ915955,blaCTX-M-27|EU916273,blaCTX-M-55|GQ456159,blaCTX-M-65|GQ456158,blaCTX-M-8|AF189721,blaDHA-1|Y16410,blaKPC-2|AY034847,blaKPC-3|HM769262,blaOXA-181|HM992946,blaSHV-105|FJ194944,blaTEM-1B|JF910132,blaTEM-52B|AF027199\n",
      "\n",
      "accuracy: 0.9281437125748503\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "          0       0.50      0.25      0.33        12\n",
      "          1       0.94      0.98      0.96       155\n",
      "\n",
      "avg / total       0.91      0.93      0.92       167\n",
      "\n",
      "TN (susceptible): 3\n",
      "FN (resistant predicted as susceptible): 3\n",
      "FP (susceptible predicted as resistant): 9\n",
      "TP (resistant): 152\n",
      "\n",
      "DRUG: phenotypic_cefepime\n",
      "\n",
      "------------------------------\n",
      "DecisionTree Set blaCTX-M-15|DQ302097\n",
      "\n",
      "accuracy: 0.6325581395348837\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "          0       0.76      0.61      0.68       136\n",
      "          1       0.50      0.67      0.57        79\n",
      "\n",
      "avg / total       0.67      0.63      0.64       215\n",
      "\n",
      "TN (susceptible): 83\n",
      "FN (resistant predicted as susceptible): 26\n",
      "FP (susceptible predicted as resistant): 53\n",
      "TP (resistant): 53\n",
      "\n",
      "DRUG: phenotypic_cefotaxime\n",
      "\n",
      "------------------------------\n",
      "DecisionTree Set blaCMY-2|X91840,blaCTX-M-14|AF252622,blaCTX-M-15|DQ302097,blaCTX-M-27|EU916273,blaKPC-2|AY034847,blaKPC-3|HM769262,blaTEM-1A|HM749966,blaTEM-1B|JF910132,blaTEM-33|GU371926,blaTEM-52B|AF027199\n",
      "\n",
      "accuracy: 0.8360655737704918\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "          0       0.75      0.25      0.38        12\n",
      "          1       0.84      0.98      0.91        49\n",
      "\n",
      "avg / total       0.82      0.84      0.80        61\n",
      "\n",
      "TN (susceptible): 3\n",
      "FN (resistant predicted as susceptible): 1\n",
      "FP (susceptible predicted as resistant): 9\n",
      "TP (resistant): 48\n",
      "\n",
      "DRUG: phenotypic_ceftazidime\n",
      "\n",
      "------------------------------\n",
      "DecisionTree Set blaCMY-2|X91840,blaCTX-M-14|AF252622,blaCTX-M-15|DQ302097,blaCTX-M-27|EU916273,blaKPC-2|AY034847,blaKPC-3|HM769262,blaOXA-1|J02967\n",
      "\n",
      "accuracy: 0.7142857142857143\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "          0       0.57      0.38      0.45        53\n",
      "          1       0.75      0.87      0.81       115\n",
      "\n",
      "avg / total       0.69      0.71      0.70       168\n",
      "\n",
      "TN (susceptible): 20\n",
      "FN (resistant predicted as susceptible): 15\n",
      "FP (susceptible predicted as resistant): 33\n",
      "TP (resistant): 100\n"
     ]
    }
   ],
   "source": [
    "# Since I don't know, use a decision tree to tell me which genes \n",
    "# and in what combinations are important for classifying resistance\n",
    "from sklearn.tree import DecisionTreeClassifier, export_graphviz\n",
    "from sklearn.feature_selection import SelectFromModel\n",
    "import graphviz\n",
    "\n",
    "accuracies = {}\n",
    "\n",
    "genes = [\"ARR-6|JF922883\",\"QnrA1|AY070235\",\"QnrB19|HM146784\",\"QnrB1|EF682133\",\n",
    "         \"QnrB20|AB379831\",\"QnrB4|DQ303921\",\"QnrB5|DQ303919\",\"QnrB6|EF523819\",\n",
    "         \"QnrS1|AB187515\",\"QnrS2|JF261185\",\"aac(3)-IIa|X51534\",\"aac(3)-IId|EU022314\",\n",
    "         \"aac(3)-IVa|X01385\",\"aac(3)-Id|AB114632\",\"aac(3)-VIa|M88012\",\n",
    "         \"aac(6)-IIc|AF162771\",\"aac(6)-Ib|M21682\",\"aac(6)Ib-cr|DQ303918\",\n",
    "         \"aac(6)Ib-cr|EF636461\",\"aacA4|KM278199\",\"aadA12|AY665771\",\"aadA13|AY713504\",\n",
    "         \"aadA16|EU675686\",\"aadA1|JQ414041\",\"aadA1|JQ480156\",\"aadA1|JX185132\",\n",
    "         \"aadA2|JQ364967\",\"aadA5|AF137361\",\"aadA7|AF224733\",\"ant(2)-Ia|M86913\",\n",
    "         \"ant(3)-Ia|X02340\",\"aph(3)-IIa|X57709\",\"aph(3)-Ia|V00359\",\"aph(3)-Ib|AF321551\",\n",
    "         \"aph(4)-Ia|V01499\",\"aph(6)-Ic|X01702\",\"aph(6)-Id|M28829\",\"armA|AY220558\",\n",
    "         \"blaCARB-2|M69058\",\"blaCMY-17|AY513266\",\"blaCMY-2|X91840\",\"blaCMY-42|HM146927\",\n",
    "         \"blaCMY-4|AF420597\",\"blaCMY-6|AJ011293\",\"blaCTX-M-14b|DQ359215\",\"blaCTX-M-14|AF252622\",\n",
    "         \"blaCTX-M-15|DQ302097\",\"blaCTX-M-1|DQ915955\",\"blaCTX-M-27|EU916273\",\n",
    "         \"blaCTX-M-55|GQ456159\",\"blaCTX-M-65|GQ456158\",\"blaCTX-M-8|AF189721\",\n",
    "         \"blaDHA-1|Y16410\",\"blaFOX-5|AY007369\",\"blaKPC-2|AY034847\",\"blaKPC-3|HM769262\",\n",
    "         \"blaKPC-4|FJ473382\",\"blaNDM-1|FN396876\",\"blaNDM-5|JN104597\",\"blaNDM-6|JN967644\",\n",
    "         \"blaNDM-7|JX262694\",\"blaOXA-181|HM992946\",\"blaOXA-1|J02967\",\"blaOXA-2|DQ310703\",\n",
    "         \"blaOXA-9|JF703130\",\"blaSHV-105|FJ194944\",\"blaSHV-12|AF462395\",\"blaSHV-2|GU064394\",\n",
    "         \"blaTEM-199|JX050178\",\"blaTEM-1A|HM749966\",\"blaTEM-1B|JF910132\",\"blaTEM-1C|FJ560503\",\n",
    "         \"blaTEM-1D|AF188200\",\"blaTEM-206|KC783461\",\"blaTEM-33|GU371926\",\"blaTEM-52B|AF027199\",\n",
    "         \"catA1|V00622\",\"catB3|AJ009818\",\"cmlA1|M64556\",\"dfrA12|AB571791\",\"dfrA14|DQ388123\",\n",
    "         \"dfrA15|HM449019\",\"dfrA17|FJ460238\",\"dfrA18|AJ310778\",\"dfrA1|JQ690541\",\"dfrA1|X00926\",\n",
    "         \"dfrA25|DQ267940\",\"dfrA29|AM237806\",\"dfrA5|X12868\",\"dfrA7|JF806498\",\"dfrA8|U10186\",\n",
    "         \"dfrB3|X72585\",\"erm(B)|AF242872\",\"erm(B)|JN899585\",\"floR|AF118107\",\"fosA7|LAPJ01000014\",\n",
    "         \"mph(A)|D16251\",\"mph(A)|U36578\",\"mph(B)|D85892\",\"mph(E)|EU294228\",\"msr(E)|EU294228\",\n",
    "         \"oqxA|EU370913\",\"oqxB|EU370913\",\"rmtB|AB103506\",\"rmtC|AB194779\",\"strA|M96392\",\"sul1|AY963803\",\n",
    "         \"sul1|CP002151\",\"sul2|FN995456\",\"sul2|GQ421466\",\"sul2|HQ840942\",\"sul3|AJ459418\",\"tet(A)|AJ517790\",\n",
    "         \"tet(A)|AY196695\",\"tet(B)|AF326777\",\"tet(B)|AP000342\",\"tet(C)|AY046276\",\"tet(C)|NC_003213\",\n",
    "         \"tet(D)|AF467077\",\"tet(G)|AF133140\",\"tet(M)|U58985\",\"tet(O)|M18896\"]\n",
    "\n",
    "genedf = amrdf[genes]\n",
    "bla = genedf.filter(regex='bla*')\n",
    "\n",
    "for drug, i_d in [['R_AMC', 0.001], ['R_AMP', 0.001], ['R_TIO', 0.001], ['R_FOX', 0.001], ['R_CRO', 0.001], ['phenotypic_aztreonam', 0.01], \n",
    "    ['phenotypic_cefazolin', 0.001], ['phenotypic_cefepime', 0.02], ['phenotypic_cefotaxime', 0.002], ['phenotypic_ceftazidime', 0.01]]: \n",
    "    y = amrdf[drug]\n",
    "    missing = y.isnull()\n",
    "\n",
    "    X = bla.values[~missing,:].astype(int)\n",
    "    y = y[~missing].astype(int)\n",
    "\n",
    "    clf = DecisionTreeClassifier(min_impurity_decrease=i_d)\n",
    "    clf.fit(X, y)\n",
    "\n",
    "    # dot_data = export_graphviz(clf, out_file=None, \n",
    "    #     feature_names=bla.columns,  \n",
    "    #     class_names=['sus','res'],  \n",
    "    #     filled=True, rounded=True,  \n",
    "    #     special_characters=True)\n",
    "    # graph = graphviz.Source(dot_data)\n",
    "    # graph\n",
    "\n",
    "    importantblagenes = bla.columns.values[clf.feature_importances_ > np.median(clf.feature_importances_)]\n",
    "    \n",
    "    impbla = np.zeros(bla.shape[0])\n",
    "    for g in importantblagenes:\n",
    "        c = bladf[g].as_matrix().astype(float)\n",
    "        impbla = np.logical_or(impbla, c)\n",
    "    y_pred = impbla.astype(int)\n",
    "    y_pred = y_pred[~missing]\n",
    "\n",
    "    print(\"\\nDRUG: \"+drug)\n",
    "    print(\"\\n------------------------------\\nDecisionTree Set {}\\n\".format(','.join(importantblagenes)))\n",
    "    accuracies[drug] = accuracy_score(y,y_pred)\n",
    "    print(\"accuracy: \"+str(accuracies[drug]))\n",
    "    print(classification_report(y, y_pred))\n",
    "    tn, fp, fn, tp = confusion_matrix(y, y_pred).ravel()\n",
    "    print(\"TN (susceptible): \"+str(tn))\n",
    "    print(\"FN (resistant predicted as susceptible): \"+str(fn))\n",
    "    print(\"FP (susceptible predicted as resistant): \"+str(fp))\n",
    "    print(\"TP (resistant): \"+str(tp))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 444,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "DRUG: R_TET\n",
      "\n",
      "------------------------------\n",
      "DecisionTree Set aph(3)-Ib|AF321551,tet(A)|AJ517790,tet(A)|AY196695,tet(B)|AF326777,tet(B)|AP000342,tet(C)|AY046276,tet(C)|NC_003213,tet(D)|AF467077,tet(G)|AF133140\n",
      "\n",
      "accuracy: 0.9849026618990863\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "          0       0.99      0.97      0.98      1052\n",
      "          1       0.98      0.99      0.99      1465\n",
      "\n",
      "avg / total       0.98      0.98      0.98      2517\n",
      "\n",
      "TN (susceptible): 1025\n",
      "FN (resistant predicted as susceptible): 11\n",
      "FP (susceptible predicted as resistant): 27\n",
      "TP (resistant): 1454\n",
      "\n",
      "DRUG: R_AZM\n",
      "\n",
      "------------------------------\n",
      "DecisionTree Set mph(A)|U36578\n",
      "\n",
      "accuracy: 0.9991126885536823\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "          0       1.00      1.00      1.00      2246\n",
      "          1       1.00      0.75      0.86         8\n",
      "\n",
      "avg / total       1.00      1.00      1.00      2254\n",
      "\n",
      "TN (susceptible): 2246\n",
      "FN (resistant predicted as susceptible): 2\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 6\n",
      "\n",
      "DRUG: R_CIP\n",
      "\n",
      "------------------------------\n",
      "DecisionTree Set QnrS2|JF261185,aac(3)-Id|AB114632,aadA1|JQ480156,blaCTX-M-14|AF252622,blaCTX-M-15|DQ302097,blaCTX-M-27|EU916273,blaCTX-M-55|GQ456159,blaKPC-2|AY034847,blaKPC-3|HM769262,catB3|AJ009818,dfrA17|FJ460238\n",
      "\n",
      "accuracy: 0.9823667711598746\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "          0       0.99      0.99      0.99      2351\n",
      "          1       0.91      0.87      0.89       201\n",
      "\n",
      "avg / total       0.98      0.98      0.98      2552\n",
      "\n",
      "TN (susceptible): 2333\n",
      "FN (resistant predicted as susceptible): 27\n",
      "FP (susceptible predicted as resistant): 18\n",
      "TP (resistant): 174\n"
     ]
    }
   ],
   "source": [
    "genedf = amrdf[genes]\n",
    "for drug in ['R_TET', 'R_AZM', 'R_CIP']: \n",
    "    y = amrdf[drug]\n",
    "    missing = y.isnull()\n",
    "\n",
    "    X = genedf.values[~missing,:].astype(int)\n",
    "    y = y[~missing].astype(int)\n",
    "\n",
    "    clf = DecisionTreeClassifier(min_impurity_decrease=0.001)\n",
    "    clf.fit(X, y)\n",
    "\n",
    "    importantgenes = genedf.columns.values[clf.feature_importances_ > np.median(clf.feature_importances_)]\n",
    "    \n",
    "    imp = np.zeros(genedf.shape[0])\n",
    "    for g in importantgenes:\n",
    "        c = genedf[g].as_matrix().astype(float)\n",
    "        imp = np.logical_or(imp, c)\n",
    "    y_pred = imp.astype(int)\n",
    "    y_pred = y_pred[~missing]\n",
    "\n",
    "    print(\"\\nDRUG: \"+drug)\n",
    "    print(\"\\n------------------------------\\nDecisionTree Set {}\\n\".format(','.join(importantgenes)))\n",
    "    accuracies[drug] = accuracy_score(y,y_pred)\n",
    "    print(\"accuracy: \"+str(accuracies[drug]))\n",
    "    print(classification_report(y, y_pred))\n",
    "    tn, fp, fn, tp = confusion_matrix(y, y_pred).ravel()\n",
    "    print(\"TN (susceptible): \"+str(tn))\n",
    "    print(\"FN (resistant predicted as susceptible): \"+str(fn))\n",
    "    print(\"FP (susceptible predicted as resistant): \"+str(fp))\n",
    "    print(\"TP (resistant): \"+str(tp))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 445,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "DRUG: R_FIS\n",
      "\n",
      "------------------------------\n",
      "DecisionTree Set aadA2|JQ364967,dfrA17|FJ460238,mph(A)|U36578,sul1|CP002151,sul2|GQ421466,sul2|HQ840942\n",
      "\n",
      "accuracy: 0.9275078369905956\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "          0       0.99      0.90      0.94      1721\n",
      "          1       0.83      0.98      0.90       831\n",
      "\n",
      "avg / total       0.94      0.93      0.93      2552\n",
      "\n",
      "TN (susceptible): 1552\n",
      "FN (resistant predicted as susceptible): 16\n",
      "FP (susceptible predicted as resistant): 169\n",
      "TP (resistant): 815\n",
      "\n",
      "DRUG: R_SXT\n",
      "\n",
      "------------------------------\n",
      "DecisionTree Set dfrA12|AB571791\n",
      "\n",
      "accuracy: 0.7351050150021432\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "          0       0.73      1.00      0.85      1691\n",
      "          1       1.00      0.04      0.07       642\n",
      "\n",
      "avg / total       0.81      0.74      0.63      2333\n",
      "\n",
      "TN (susceptible): 1691\n",
      "FN (resistant predicted as susceptible): 618\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 24\n"
     ]
    }
   ],
   "source": [
    "genedf = amrdf[genes]\n",
    "for drug,i_d in [['R_FIS', 0.01],['R_SXT', 0.01]]: \n",
    "    y = amrdf[drug]\n",
    "    missing = y.isnull()\n",
    "\n",
    "    X = genedf.values[~missing,:].astype(int)\n",
    "    y = y[~missing].astype(int)\n",
    "\n",
    "    clf = DecisionTreeClassifier(min_impurity_decrease=i_d)\n",
    "    clf.fit(X, y)\n",
    "\n",
    "    importantgenes = genedf.columns.values[clf.feature_importances_ > np.median(clf.feature_importances_)]\n",
    "    \n",
    "    imp = np.zeros(genedf.shape[0])\n",
    "    for g in importantgenes:\n",
    "        c = genedf[g].as_matrix().astype(float)\n",
    "        imp = np.logical_or(imp, c)\n",
    "    y_pred = imp.astype(int)\n",
    "    y_pred = y_pred[~missing]\n",
    "\n",
    "    print(\"\\nDRUG: \"+drug)\n",
    "    print(\"\\n------------------------------\\nDecisionTree Set {}\\n\".format(','.join(importantgenes)))\n",
    "    accuracies[drug] = accuracy_score(y,y_pred)\n",
    "    print(\"accuracy: \"+str(accuracies[drug]))\n",
    "    print(classification_report(y, y_pred))\n",
    "    tn, fp, fn, tp = confusion_matrix(y, y_pred).ravel()\n",
    "    print(\"TN (susceptible): \"+str(tn))\n",
    "    print(\"FN (resistant predicted as susceptible): \"+str(fn))\n",
    "    print(\"FP (susceptible predicted as resistant): \"+str(fp))\n",
    "    print(\"TP (resistant): \"+str(tp))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 446,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "DRUG: R_CIP\n",
      "\n",
      "------------------------------\n",
      "DecisionTree Set blaCTX-M-15|DQ302097,dfrA17|FJ460238\n",
      "\n",
      "accuracy: 0.9721786833855799\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "          0       0.98      0.99      0.99      2351\n",
      "          1       0.92      0.71      0.80       201\n",
      "\n",
      "avg / total       0.97      0.97      0.97      2552\n",
      "\n",
      "TN (susceptible): 2338\n",
      "FN (resistant predicted as susceptible): 58\n",
      "FP (susceptible predicted as resistant): 13\n",
      "TP (resistant): 143\n",
      "\n",
      "DRUG: R_NAL\n",
      "\n",
      "------------------------------\n",
      "DecisionTree Set \n",
      "\n",
      "accuracy: 0.9802828975567939\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "          0       0.98      1.00      0.99      2287\n",
      "          1       0.00      0.00      0.00        46\n",
      "\n",
      "avg / total       0.96      0.98      0.97      2333\n",
      "\n",
      "TN (susceptible): 2287\n",
      "FN (resistant predicted as susceptible): 46\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 0\n",
      "\n",
      "DRUG: phenotypic_levofloxacin\n",
      "\n",
      "------------------------------\n",
      "DecisionTree Set aac(6)Ib-cr|DQ303918,blaCMY-2|X91840,blaTEM-1B|JF910132,dfrA12|AB571791,tet(D)|AF467077\n",
      "\n",
      "accuracy: 0.5981735159817352\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "          0       0.20      0.27      0.23        48\n",
      "          1       0.77      0.69      0.73       171\n",
      "\n",
      "avg / total       0.65      0.60      0.62       219\n",
      "\n",
      "TN (susceptible): 13\n",
      "FN (resistant predicted as susceptible): 53\n",
      "FP (susceptible predicted as resistant): 35\n",
      "TP (resistant): 118\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/matt/anaconda3/envs/phenores4/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.\n",
      "  'precision', 'predicted', average, warn_for)\n"
     ]
    }
   ],
   "source": [
    "genedf = amrdf[genes]\n",
    "# These are determined by chromosomal mutations in the \"Quinolone resistance-determining regions\"\n",
    "for drug,i_d in [['R_CIP', 0.01],['R_NAL', 0.01], ['phenotypic_levofloxacin',0.015]]: \n",
    "    y = amrdf[drug]\n",
    "    missing = y.isnull()\n",
    "\n",
    "    X = genedf.values[~missing,:].astype(int)\n",
    "    y = y[~missing].astype(int)\n",
    "\n",
    "    clf = DecisionTreeClassifier(min_impurity_decrease=i_d)\n",
    "    clf.fit(X, y)\n",
    "\n",
    "    importantgenes = genedf.columns.values[clf.feature_importances_ > np.median(clf.feature_importances_)]\n",
    "    \n",
    "    imp = np.zeros(genedf.shape[0])\n",
    "    for g in importantgenes:\n",
    "        c = genedf[g].as_matrix().astype(float)\n",
    "        imp = np.logical_or(imp, c)\n",
    "    y_pred = imp.astype(int)\n",
    "    y_pred = y_pred[~missing]\n",
    "\n",
    "    print(\"\\nDRUG: \"+drug)\n",
    "    print(\"\\n------------------------------\\nDecisionTree Set {}\\n\".format(','.join(importantgenes)))\n",
    "    accuracies[drug] = accuracy_score(y,y_pred)\n",
    "    print(\"accuracy: \"+str(accuracies[drug]))\n",
    "    print(classification_report(y, y_pred))\n",
    "    tn, fp, fn, tp = confusion_matrix(y, y_pred).ravel()\n",
    "    print(\"TN (susceptible): \"+str(tn))\n",
    "    print(\"FN (resistant predicted as susceptible): \"+str(fn))\n",
    "    print(\"FP (susceptible predicted as resistant): \"+str(fp))\n",
    "    print(\"TP (resistant): \"+str(tp))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 428,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "DRUG: R_GEN\n",
      "\n",
      "------------------------------\n",
      "DecisionTree Set aac(3)-IIa|X51534,aac(3)-IId|EU022314,aac(3)-VIa|M88012,ant(2)-Ia|M86913,ant(3)-Ia|X02340\n",
      "\n",
      "accuracy: 0.9760971786833855\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "          0       0.99      0.99      0.99      2210\n",
      "          1       0.91      0.92      0.91       342\n",
      "\n",
      "avg / total       0.98      0.98      0.98      2552\n",
      "\n",
      "TN (susceptible): 2178\n",
      "FN (resistant predicted as susceptible): 29\n",
      "FP (susceptible predicted as resistant): 32\n",
      "TP (resistant): 313\n",
      "\n",
      "DRUG: phenotypic_streptomycin\n",
      "\n",
      "------------------------------\n",
      "DecisionTree Set aac(6)-Ib|M21682,aadA2|JQ364967,ant(3)-Ia|X02340,aph(6)-Id|M28829,strA|M96392\n",
      "\n",
      "accuracy: 0.9480796586059744\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "          0       0.97      0.94      0.96       861\n",
      "          1       0.92      0.95      0.93       545\n",
      "\n",
      "avg / total       0.95      0.95      0.95      1406\n",
      "\n",
      "TN (susceptible): 813\n",
      "FN (resistant predicted as susceptible): 25\n",
      "FP (susceptible predicted as resistant): 48\n",
      "TP (resistant): 520\n",
      "\n",
      "DRUG: phenotypic_kanamycin\n",
      "\n",
      "------------------------------\n",
      "DecisionTree Set aph(3)-IIa|X57709,aph(3)-Ia|V00359\n",
      "\n",
      "accuracy: 0.985494106980961\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "          0       0.98      1.00      0.99       995\n",
      "          1       1.00      0.85      0.92       108\n",
      "\n",
      "avg / total       0.99      0.99      0.98      1103\n",
      "\n",
      "TN (susceptible): 995\n",
      "FN (resistant predicted as susceptible): 16\n",
      "FP (susceptible predicted as resistant): 0\n",
      "TP (resistant): 92\n"
     ]
    }
   ],
   "source": [
    "genedf = amrdf[genes]\n",
    "for drug,i_d in [['R_GEN', 0.01], ['phenotypic_streptomycin',0.008], ['phenotypic_kanamycin',0.008]]: \n",
    "    y = amrdf[drug]\n",
    "    missing = y.isnull()\n",
    "\n",
    "    X = genedf.values[~missing,:].astype(int)\n",
    "    y = y[~missing].astype(int)\n",
    "\n",
    "    clf = DecisionTreeClassifier(min_impurity_decrease=i_d)\n",
    "    clf.fit(X, y)\n",
    "\n",
    "    importantgenes = genedf.columns.values[clf.feature_importances_ > np.median(clf.feature_importances_)]\n",
    "    \n",
    "    imp = np.zeros(genedf.shape[0])\n",
    "    for g in importantgenes:\n",
    "        c = genedf[g].as_matrix().astype(float)\n",
    "        imp = np.logical_or(imp, c)\n",
    "    y_pred = imp.astype(int)\n",
    "    y_pred = y_pred[~missing]\n",
    "\n",
    "    print(\"\\nDRUG: \"+drug)\n",
    "    print(\"\\n------------------------------\\nDecisionTree Set {}\\n\".format(','.join(importantgenes)))\n",
    "    print(\"accuracy: \"+str(accuracy_score(y,y_pred)))\n",
    "    print(classification_report(y, y_pred))\n",
    "    tn, fp, fn, tp = confusion_matrix(y, y_pred).ravel()\n",
    "    print(\"TN (susceptible): \"+str(tn))\n",
    "    print(\"FN (resistant predicted as susceptible): \"+str(fn))\n",
    "    print(\"FP (susceptible predicted as resistant): \"+str(fp))\n",
    "    print(\"TP (resistant): \"+str(tp))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 447,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "DRUG: phenotypic_ertapenem\n",
      "\n",
      "------------------------------\n",
      "DecisionTree Set blaKPC-2|AY034847,blaKPC-3|HM769262,blaNDM-5|JN104597,strA|M96392\n",
      "\n",
      "accuracy: 0.8935185185185185\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "          0       0.89      0.99      0.94       177\n",
      "          1       0.90      0.46      0.61        39\n",
      "\n",
      "avg / total       0.89      0.89      0.88       216\n",
      "\n",
      "TN (susceptible): 175\n",
      "FN (resistant predicted as susceptible): 21\n",
      "FP (susceptible predicted as resistant): 2\n",
      "TP (resistant): 18\n",
      "\n",
      "DRUG: phenotypic_meropenem\n",
      "\n",
      "------------------------------\n",
      "DecisionTree Set blaNDM-5|JN104597,blaNDM-6|JN967644,blaNDM-7|JX262694,cmlA1|M64556,strA|M96392,tet(A)|AY196695\n",
      "\n",
      "accuracy: 0.9315068493150684\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "          0       0.95      0.98      0.96       194\n",
      "          1       0.78      0.56      0.65        25\n",
      "\n",
      "avg / total       0.93      0.93      0.93       219\n",
      "\n",
      "TN (susceptible): 190\n",
      "FN (resistant predicted as susceptible): 11\n",
      "FP (susceptible predicted as resistant): 4\n",
      "TP (resistant): 14\n",
      "\n",
      "DRUG: phenotypic_doripenem\n",
      "\n",
      "------------------------------\n",
      "DecisionTree Set aac(3)-IId|EU022314,aadA2|JQ364967,blaCMY-42|HM146927,blaKPC-2|AY034847,blaNDM-7|JX262694,dfrA5|X12868,strA|M96392,tet(B)|AF326777\n",
      "\n",
      "accuracy: 0.6666666666666666\n",
      "             precision    recall  f1-score   support\n",
      "\n",
      "          0       0.89      0.66      0.76        47\n",
      "          1       0.36      0.69      0.47        13\n",
      "\n",
      "avg / total       0.77      0.67      0.69        60\n",
      "\n",
      "TN (susceptible): 31\n",
      "FN (resistant predicted as susceptible): 4\n",
      "FP (susceptible predicted as resistant): 16\n",
      "TP (resistant): 9\n"
     ]
    }
   ],
   "source": [
    "genedf = amrdf[genes]\n",
    "for drug,i_d in [['phenotypic_ertapenem', 0.01], ['phenotypic_meropenem',0.008], ['phenotypic_doripenem',0.01]]: \n",
    "    y = amrdf[drug]\n",
    "    missing = y.isnull()\n",
    "\n",
    "    X = genedf.values[~missing,:].astype(int)\n",
    "    y = y[~missing].astype(int)\n",
    "\n",
    "    clf = DecisionTreeClassifier(min_impurity_decrease=i_d)\n",
    "    clf.fit(X, y)\n",
    "\n",
    "    importantgenes = genedf.columns.values[clf.feature_importances_ > np.median(clf.feature_importances_)]\n",
    "    \n",
    "    imp = np.zeros(genedf.shape[0])\n",
    "    for g in importantgenes:\n",
    "        c = genedf[g].as_matrix().astype(float)\n",
    "        imp = np.logical_or(imp, c)\n",
    "    y_pred = imp.astype(int)\n",
    "    y_pred = y_pred[~missing]\n",
    "\n",
    "    print(\"\\nDRUG: \"+drug)\n",
    "    print(\"\\n------------------------------\\nDecisionTree Set {}\\n\".format(','.join(importantgenes)))\n",
    "    accuracies[drug] = accuracy_score(y,y_pred)\n",
    "    print(\"accuracy: \"+str(accuracies[drug]))\n",
    "    print(classification_report(y, y_pred))\n",
    "    tn, fp, fn, tp = confusion_matrix(y, y_pred).ravel()\n",
    "    print(\"TN (susceptible): \"+str(tn))\n",
    "    print(\"FN (resistant predicted as susceptible): \"+str(fn))\n",
    "    print(\"FP (susceptible predicted as resistant): \"+str(fp))\n",
    "    print(\"TP (resistant): \"+str(tp))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 385,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "phenotypic_amikacin                            13.0\n",
       "phenotypic_amoxicillin-clavulanic acid        364.0\n",
       "phenotypic_ampicillin                         707.0\n",
       "phenotypic_ampicillin-sulbactam                49.0\n",
       "phenotypic_azithromycin                         1.0\n",
       "phenotypic_aztreonam                           92.0\n",
       "phenotypic_cefalexin                            1.0\n",
       "phenotypic_cefazolin                          155.0\n",
       "phenotypic_cefepime                            79.0\n",
       "phenotypic_cefotaxime                          49.0\n",
       "phenotypic_cefotaxime-clavulanic acid           1.0\n",
       "phenotypic_cefotetan                            0.0\n",
       "phenotypic_cefoxitin                          294.0\n",
       "phenotypic_ceftazidime                        115.0\n",
       "phenotypic_ceftazidime-clavulanic acid          0.0\n",
       "phenotypic_ceftiofur                          276.0\n",
       "phenotypic_ceftriaxone                        479.0\n",
       "phenotypic_chloramphenicol                     94.0\n",
       "phenotypic_ciprofloxacin                      180.0\n",
       "phenotypic_colistin                             0.0\n",
       "phenotypic_doripenem                           13.0\n",
       "phenotypic_ertapenem                           39.0\n",
       "phenotypic_fosfomycin                           1.0\n",
       "phenotypic_gentamicin                         249.0\n",
       "phenotypic_imipenem                            18.0\n",
       "phenotypic_kanamycin                          108.0\n",
       "phenotypic_levofloxacin                       171.0\n",
       "phenotypic_meropenem                           25.0\n",
       "phenotypic_minocycline                          0.0\n",
       "phenotypic_nalidixic acid                      24.0\n",
       "                                              ...  \n",
       "phenotypic_cefotetan.1                          0.0\n",
       "phenotypic_cefoxitin.1                          0.0\n",
       "phenotypic_ceftazidime.1                        0.0\n",
       "phenotypic_ceftazidime-clavulanic acid.1        0.0\n",
       "phenotypic_ceftiofur.1                          0.0\n",
       "phenotypic_ceftriaxone.1                        0.0\n",
       "phenotypic_chloramphenicol.1                    0.0\n",
       "phenotypic_ciprofloxacin.1                      0.0\n",
       "phenotypic_colistin.1                           0.0\n",
       "phenotypic_doripenem.1                          0.0\n",
       "phenotypic_ertapenem.1                          0.0\n",
       "phenotypic_fosfomycin.1                         0.0\n",
       "phenotypic_gentamicin.1                         0.0\n",
       "phenotypic_imipenem.1                           0.0\n",
       "phenotypic_kanamycin.1                          0.0\n",
       "phenotypic_levofloxacin.1                       0.0\n",
       "phenotypic_meropenem.1                          0.0\n",
       "phenotypic_minocycline.1                        0.0\n",
       "phenotypic_nalidixic acid.1                     0.0\n",
       "phenotypic_nitrofurantoin.1                     0.0\n",
       "phenotypic_piperacillin.1                       0.0\n",
       "phenotypic_piperacillin-tazobactam.1            0.0\n",
       "phenotypic_polymyxin B.1                        0.0\n",
       "phenotypic_streptomycin.1                       0.0\n",
       "phenotypic_sulbactam.1                          0.0\n",
       "phenotypic_sulfisoxazole.1                      0.0\n",
       "phenotypic_tetracycline.1                       0.0\n",
       "phenotypic_tigecycline.1                        0.0\n",
       "phenotypic_tobramycin.1                         0.0\n",
       "phenotypic_trimethoprim-sulfamethoxazole.1      0.0\n",
       "Length: 82, dtype: float64"
      ]
     },
     "execution_count": 385,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdf = amrdf.filter(regex=\"phenotypic_*\")\n",
    "pdf.sum(axis=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 472,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "levofloxacin     0.5982\n",
      "cefepime         0.6326\n",
      "doripenem        0.6667\n",
      "ceftazidime      0.7143\n",
      "SXT              0.7351\n",
      "cefotaxime       0.8361\n",
      "aztreonam        0.8584\n",
      "ertapenem        0.8935\n",
      "FIS              0.9275\n",
      "cefazolin        0.9281\n",
      "meropenem        0.9315\n",
      "FOX              0.9500\n",
      "AMC              0.9670\n",
      "CIP              0.9722\n",
      "NAL              0.9803\n",
      "AMP              0.9816\n",
      "TET              0.9849\n",
      "CRO              0.9898\n",
      "TIO              0.9923\n",
      "AZM              0.9991\n"
     ]
    }
   ],
   "source": [
    "for key, value in sorted(accuracies.items(), key=lambda item: (item[1],item[0])):\n",
    "    key=key.replace('phenotypic_','')\n",
    "    key=key.replace('R_','')\n",
    "    print(\"{key: <12} {value:10.4f}\".format(key=key,value=value))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
